compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp

```diff
@@ -309,12 +356,25 @@ struct LLVMGPUTileAndDistributePass
       switch (promoteSharedMemPattern) {
         case GPUPromoteSharedMemPattern::ContractionOpPattern:
           populatePromotionPatterns(context, promotionPatterns,
-                                    promoteSharedMemPattern, {0, 1});
+                                    contractOpFilter, {0, 1});
 
           break;
         case GPUPromoteSharedMemPattern::TransposeOpPattern:
-          populatePromotionPatterns(context, promotionPatterns,
-                                    promoteSharedMemPattern, {0});
+          funcOp.walk(
+              [&context, &promotionPatterns](linalg::GenericOp linalgOp) {
+                linalgOp.dump();
+                // For each linalg generic, determine out its operands to
+                // promote, then create a filter that will only apply to it's
+                // configuration.
+                SmallVector<int64_t> operandsToPromote =
+                    getTransposedOperands(linalgOp);
+                populatePromotionPatterns(
+                    context, promotionPatterns,
+                    [operandsToPromote](Operation *op) -> LogicalResult {
+                      return transposeFilter(op, operandsToPromote);
+                    },
+                    operandsToPromote);
```

Comment:
it sounds a bit convoluted, since we are looping through all the ops we may as well make the filter just check that this is the op. Also it would be good to add a comment explaining that this is a workaround because transpose operands should be a lambda and that this will go away once we move away from pattern rewrites.

Reply:
added more description. let me know if its not clear.